File: leetcode/strings/easy/java_solution/ReverseVowelsOfString_345.java
package leetcode.strings.easy.java_solution;

import java.util.LinkedList;

public class ReverseVowelsOfString_345 {

    private static boolean isVowel(char c) {
        return c == 'a' || c == 'i' || c == 'e' || c == 'o' || c == 'u' || c == 'A' || c == 'I' || c == 'E' || c == 'O' || c == 'U';
    }

    static class SolutionTwoPointers {
        public String reverseVowels(String s) {
            char[] word = s.toCharArray();

            int left = 0, right = word.length - 1;

            while (left < right) {
                if (!isVowel(word[left])) {
                    left++;
                } else if (!isVowel(word[right])) {
                    right--;
                } else {
                    char temp = word[left];
                    word[left] = word[right];
                    word[right] = temp;
                    left++;
                    right--;
                }
            }
            return new String(word);
        }
    }

    static class Solution {
        public String reverseVowels(String s) {
            final char[] word = s.toCharArray();
            final LinkedList<Character> vowels = new LinkedList<>();

            for (char current : word) {
                if (isVowel(current)) {
                    vowels.add(current);
                }
            }
            for (int i = 0; i < word.length; i++) {
                if (isVowel(word[i])) {
                    word[i] = vowels.removeLast();
                }
            }
            return new String(word);
        }
    }
}

File: leetcode/strings/easy/golang_solution/reverse_vowels_of_string_345.go
package golang_solution

func reverseVowelsTwoPointers(s string) string {
	word := []rune(s)

	left, right := 0, len(word)-1

	for left < right {
		if !isVowel(word[left]) {
			left++
		} else if !isVowel(word[right]) {
			right--
		} else {
			word[left], word[right] = word[right], word[left]
			left++
			right--
		}
	}
	return string(word)
}

func reverseVowels(s string) string {
	vowels := make([]rune, 0)
	runes := []rune(s)

	for _, v := range runes {
		if isVowel(v) {
			vowels = append(vowels, v)
		}
	}
	right := len(vowels) - 1
	for i, v := range runes {
		if isVowel(v) {
			runes[i] = vowels[right]
			right--
		}
	}
	return string(runes)
}

func isVowel(c rune) bool {
	return c == 'a' || c == 'i' || c == 'e' || c == 'o' || c == 'u' || c == 'A' || c == 'I' || c == 'E' || c == 'O' || c == 'U'
}
